src/ImageRunModal.jsx

```diff
@@ -27,13 +28,109 @@ const units = {
     },
 };
 
+class PublishPortsForm extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            publish: [{ key: 0, containerPort: null, hostPort: null, protocol: 'TCP' }],
+            keyCounter: 1,
```

Comment:
It seems a little odd to keep the counter inside the state, instead of a simple `this.keyCounter` object property. But you always update it together with `publish`, so it's ok if you prefer this.